node/uhpo/src/payout_update.rs

```diff
@@ -0,0 +1,275 @@
+use std::fmt::Error;
+
+use bitcoin::{
+    absolute::LockTime,
+    hashes::Hash,
+    key::{Keypair, Secp256k1},
+    secp256k1::{Message, Scalar, SecretKey},
+    sighash::{Prevouts, SighashCache},
+    transaction::Version,
+    Address, Amount, OutPoint, ScriptBuf, Sequence, TapSighashType, Transaction, TxIn, TxOut,
+    Witness,
+};
+
+use crate::error::UhpoError;
+
+pub struct PayoutUpdate {
+    transaction: Transaction,
+    coinbase_txout: TxOut,
+    prev_update_txout: Option<TxOut>,
+}
+
+impl PayoutUpdate {
+    pub fn new(
+        prev_update_tx: Option<Transaction>,
+        coinbase_tx: Transaction,
+        next_out_address: Address,
+        projected_fee: Amount,
+    ) -> Result<Self, Error> {
+        let prev_update_txout = prev_update_tx.as_ref().map(|tx| tx.output[0].clone());
+        let coinbase_txout = coinbase_tx.output[0].clone();
+
+        let payout_update_tx = build_transaction(
+            &coinbase_tx,
+            prev_update_tx.as_ref(),
+            next_out_address,
+            projected_fee,
+            &coinbase_txout,
+            prev_update_txout.as_ref(),
+        )?;
+
+        Ok(PayoutUpdate {
+            transaction: payout_update_tx,
+            coinbase_txout,
+            prev_update_txout,
+        })
+    }
+
+    pub fn add_coinbase_sig(
+        &mut self,
+        private_key: &SecretKey,
+        tweak: &Option<Scalar>,
+    ) -> Result<(), UhpoError> {
+        let prevouts = match &self.prev_update_txout {
+            Some(prev_update_txout) => vec![&self.coinbase_txout, prev_update_txout],
+            None => vec![&self.coinbase_txout],
+        };
+
+        add_signature(&mut self.transaction, 0, &prevouts, private_key, tweak)
+    }
+
+    pub fn add_prev_update_sig(
+        &mut self,
+        private_key: &SecretKey,
+        tweak: &Option<Scalar>,
+    ) -> Result<(), UhpoError> {
+        let prev_update_txout = self
+            .prev_update_txout
+            .as_ref()
+            .ok_or(UhpoError::NoPrevUpdateTxOut)?;
+        let prevouts = vec![&self.coinbase_txout, prev_update_txout];
+
+        add_signature(&mut self.transaction, 1, &prevouts, private_key, tweak)
+    }
+
+    pub fn build(self) -> Transaction {
+        self.transaction
+    }
+}
+
+fn build_transaction(
+    coinbase_tx: &Transaction,
+    prev_update_tx: Option<&Transaction>,
+    next_out_address: Address,
+    projected_fee: Amount,
+    coinbase_txout: &TxOut,
+    prev_update_txout: Option<&TxOut>,
```

Comment:
We don't need any of these arguments to be a reference.

Here's a working patch that shows how. https://github.com/pool2win/braidpool/commit/638d3c74d6e6695b535f241985ce3de010ed756b